### ISSUE TITLE
bug: datetime intersection observer not firing in modal

### ISSUE THREAD

**Issue by enriqueloz88** [ionic-team/ionic-framework#23992]
### Prequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

- [ ] v4.x
- [ ] v5.x
- [X] v6.x

### Current Behavior

The datetime component works really fine when I used in the page directly. On the contrary, when I try to open it with an ion-modal the component broke. The datetime seems to display correctly in the modal, and when I select a date of the same month as the current one it works correctly. However, as soon as I want to change the month using the top buttons or the scroll, the component does not update the months in the upper part of the date-picker (in addition, it only lets you advance or go back one month with respect to the current one, both through the upper buttons and through scroll). However, when you start punching on any date the component only advances to the next month.

See the video for a better compression of the problem (sorry for the low quality). In the first seconds when dates relative to the current month were selected it worked, however, as soon as you want to change the month strange behaviors appear:

https://user-images.githubusercontent.com/61509169/135252509-261918c7-d710-4c11-844d-65aa9b760632.mp4

I tried with diffent widths and heigths but I can't resolve the issue.

### Expected Behavior

The datepicker should work the same way inside an ion-modal / popover as it does outside of them.

### Steps to Reproduce

- Create a project with the beta 6 of Ionic v6
- Create a container that opens the modal

```
      <div (click)="openDateModal()" class="container"> Open Date Modal  </div>
```
- Create the corresponding function to open it
```
  dateModal: HTMLIonModalElement;

  async openDateModal() {
    this.dateModal = await this.modalCtrl.create({
      component: DatePickerPage,
      cssClass: 'modal-with-transparency',
    });

    this.dateModal.onDidDismiss().then((data) => {
      this.selectedTime = data.data;
    });

    return this.dateModal.present();
  }
```
- Create a new page with the following code, that will contain the datepicker:
__html__
```
<div id="background" (click)="dismiss($event)">
  <ion-datetime id="picker" [(ngModel)]="date">
    <ion-buttons slot="buttons">
      <ion-button (click)="confirm()">Done!</ion-button>
    </ion-buttons>
  </ion-datetime>
</div>
```

__css__
```
ion-datetime {
  height: min(60%, 500px);
  width: min(85%, 340px);
}

#background {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
}
```

__ts__
```
  date: any;
  constructor(private modalCtrl: ModalController) {}

  ngOnInit() {}

  dismiss(e) {
    if (e.target.id == 'background') {
      this.modalCtrl.dismiss();
    }
  }

  confirm() {
    const datetime = <any>document.querySelector('#picker');
    datetime.confirm().then((res) => {
      this.modalCtrl.dismiss(new Date(this.date));
    });
  }
```

### Ionic Info

Ionic:

   Ionic CLI                     : 6.17.1 (C:\Users\kikel\AppData\Roaming\npm\node_modules\@ionic\cli)
   Ionic Framework               : @ionic/angular 6.0.0-beta.6
   @angular-devkit/build-angular : 12.1.4
   @angular-devkit/schematics    : 12.1.4
   @angular/cli                  : 12.1.4
   @ionic/angular-toolkit        : 4.0.0

Capacitor:

   Capacitor CLI      : 3.2.3
   @capacitor/android : 3.2.3
   @capacitor/core    : 3.2.3
   @capacitor/ios     : not installed

Utility:

   cordova-res : 0.15.3
   native-run  : 1.4.1

System:

   NodeJS : v14.15.4 (C:\Program Files\nodejs\node.exe)
   npm    : 6.14.10
   OS     : Windows 10

**Comment by enriqueloz88** [comment]
Ok, it seems to be a problem related to issue #23985. Adjusting the size of the datepicker to a height of 100% make it works correctly. Anyway, it should still work in the same way for smaller sizes like the datepicker of the video.

**Comment by liamdebeasi** [comment]
Thanks for the issue. Are you able to reproduce this in an Ionic starter app and provide a link to the repo? I am not able to reproduce this on my end. Your video does not appear zoomed in, so I am not sure it is exactly the same issue as #23985.

**Comment by enriqueloz88** [comment]
I created a new repo with the issue. Here is the link:

https://github.com/enriqueloz88/datepicker-bug

The problem appears in some devices like the Pixel 2XL or the iPhone 5/SE in the devtools (also when you create the native app via android studio or xcode). In some other devices like the iPhone X the problem is not visible. Therebefore, it seems that it is dependent on the width and height of the component, and if you try to modify these parameters at the end you make it work, but it is too random and I have not yet managed to make it work for all devices.

**Comment by enriqueloz88** [comment]
I have made a new video with the problem just in case. At the beginning of the video I ran the application from the repository that I passed previously on the Pixel, and the component fails. However, when reloading the page, if we put the dimensions of another phone like the iPhone X, it does work

https://user-images.githubusercontent.com/61509169/135326598-e13c9fc3-a1f2-4cb3-8b9c-d405a8daa649.mp4

**Comment by liamdebeasi** [comment]
Thanks! I can reproduce this behavior.

**Comment by sean-perkins** [comment]
Hello :wave: following up on this issue.

The problem here is that when unspecified the `ion-datetime` will auto size to its container. For different devices, this will be a different width.

For "Pixel 5" emulation, the `ion-datetime` width is set to `334.05px`. 

The odd fractional value is affecting the intersection observers ability to handle thresholds correctly. We can replicate the issue with all usages of `ion-datetime` in a modal, if we set the width explicitly to `334.05px`. 

If we set the value to an integer whole number, i.e. `334px` the problem goes away.

Obviously this isn't ideal for auto-sizing layouts, where you aren't specifying the end width of the datetime. I have an adjustment to the intersection observer usage that can account for this internally.
